rust/src/utils/kmeans.rs

```diff
@@ -0,0 +1,306 @@
+        data: Arc<FixedSizeListArray>,
+        k: u32,
+        params: &KMeansParams,
+    ) -> Self {
+        let mut best_kmeans = KMeans::empty(k, data.value_length());
+        let mut best_stddev = f32::MAX;
+
+        for _ in 1..=params.redos {
+            let mut kmeans = kmean_plusplus(data.clone(), k, &mut rand::thread_rng()).await;
+            let mut last_membership = kmeans.compute_membership(data.clone()).await;
+            for i in 1..=params.max_iters {
+                let new_kmeans = last_membership.to_kmean().await.unwrap();
+                let new_membership = new_kmeans.compute_membership(data.clone()).await;
+                if (new_membership.distance_sum() - last_membership.distance_sum()).abs()
+                    / last_membership.distance_sum()
+                    < params.tolerance
+                {
+                    kmeans = new_kmeans;
+                    last_membership = new_membership;
+                    #[cfg(debug_assertions)]
+                    println!("Kmeans: early converged at iteration: {i}");
+                    break;
+                }
+                kmeans = new_kmeans;
+                last_membership = new_membership;
+            }
+            // Optimize for balanced clusters instead of minimal distance.
+            let stddev = last_membership.hist_stddev();
+            if stddev < best_stddev {
+                best_kmeans = kmeans;
+                best_stddev = stddev;
+            }
+        }
+
+        best_kmeans
+    }
+
+    async fn compute_membership(&self, data: Arc<FixedSizeListArray>) -> KMeanMembership {
+        let cluster_with_distances = stream::iter(0..data.len())
+            // make tiles of input data to split between threads.
+            .chunks(1024)
+            .zip(repeat_with(|| (data.clone(), self.centroids.clone())))
+            .map(|(indices, (data, centroids))| async move {
+                let data = tokio::task::spawn_blocking(move || {
+                    let mut results = vec![];
+                    for idx in indices {
+                        let value_arr = data.value(idx);
+                        let vector: &Float32Array = as_primitive_array(&value_arr);
+                        let distances = l2_distance(vector, centroids.as_ref()).unwrap();
+                        let cluster_id = argmin(distances.as_ref()).unwrap();
+                        let distance = distances.value(cluster_id as usize);
+                        results.push((cluster_id, distance))
+                    }
+                    results
+                })
+                .await?;
+                Ok::<Vec<_>, Error>(data)
+            })
+            .buffered(num_cpus::get())
+            .try_collect::<Vec<_>>()
+            .await
+            .unwrap();
+        let k = self.k;
+        KMeanMembership {
+            data,
+            cluster_ids: cluster_with_distances
+                .iter()
+                .flatten()
```

Comment:
Is this equivalent to flatten? Or is that not available here

Reply:
this is flatten()?

Reply:
Oh I saw flat_map